backend/auth/middleware.go

```diff
@@ -117,6 +148,50 @@ func (am *AuthMiddleware) AuthenticateRequest(next http.Handler) http.Handler {
 	})
 }
 
+func (am *AuthMiddleware) verifyState(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logger := klog.FromContext(r.Context())
+
+		authCtx := GetAuthContext(r.Context())
+		if authCtx.SessionState == nil {
+			logger.V(2).Info("No session state found in auth context")
+			writeErrorResponse(w, http.StatusUnauthorized, kubebindv1alpha2.ErrorCodeAuthenticationFailed, "Authentication required", "No valid session found")
+			return
+		}
+
+		state := authCtx.SessionState
+		// Validate session fields are present
+		switch {
+		case state.Token.Subject == "" || state.Token.Issuer == "" || state.SessionID == "":
+			logger.V(2).Info("Invalid session state: missing required fields")
+			writeErrorResponse(w, http.StatusUnauthorized, kubebindv1alpha2.ErrorCodeAuthenticationFailed, "Authentication required", "Invalid session state: missing required fields")
+			return
+		case state.IsExpired():
+			logger.V(2).Info("Session expired", "sessionID", state.SessionID)
+			writeErrorResponse(w, http.StatusUnauthorized, kubebindv1alpha2.ErrorCodeAuthenticationFailed, "Authentication required", "Session has expired")
+			return
+		case !am.isValidSession(state.SessionID):
+			logger.V(2).Info("Session ID not found or expired", "sessionID", state.SessionID)
+			writeErrorResponse(w, http.StatusUnauthorized, kubebindv1alpha2.ErrorCodeAuthenticationFailed, "Authentication required", "Session ID not found or expired")
+			return
+		default:
+			// Session is valid
+			authCtx.IsValid = true
+		}
+		ctx := context.WithValue(r.Context(), AuthContextKey, authCtx)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// isValidSession checks if a session ID exists and hasn't expired
+func (am *AuthMiddleware) isValidSession(sessionID string) bool {
+	sessionInfo, err := am.sessionStore.Load(sessionID)
+	if err != nil {
+		return false
+	}
+	return time.Now().Before(sessionInfo.ExpiresAt)
```

Comment:
What would be the consequences if `time.Now().Before(sessionInfo.ExpiresAt)` returns `true` at this very moment, but the request finishes already at `time.Now() > sessionInfo.ExpiresAt`? Not sure what the consensus is on these time-based checks. I believe I've seen some implementations with a buffer period where `should_renew_at > valid_since && should_renew_at < expires_at`. I'm just curious how this would be handled rather than saying it needs to change because I'm not sure what the consequences would be :)

The same is true for `State.IsExpired()`.

Reply:
Ah, the renewal flow is not yet in place? Okay, probably not much more we can do here at the moment. Thanks for explaining!